Documentation for 3.0.0 updates (#252)
alleyinteractive/irving: stories/05-architecture/01-Configuration.stories.mdx

* **array**: configuration expects either an array or a function returning an array. If you provide an array, this array will be concatenated to Irving core's default configuration. If you provide a function, Irving core's default configuration will be passed to this function as its first parameter and expect you to return a modified version of that array. This configuration will be used as-is, so all mutations or merging operations are up to you to perform.
* **string**: configuration expects a string, and any string you provide will replace Irving core's value.
* **function**: configuration expects a function, and any function you provide will either replace Irving core's version of the same function or be run in sequence, passing the return value from the previous function to each subsequent function. Configuration descriptions below will note which of these cases apply.

Irving will expect you to use one of two methods for configuration depending on what you'd like to configure:

> * **filesystem**: used primarly for server-side configurations outside the webpack build. Place the configuration file a the specified **path** and Irving will find your configuration.

Reviewer: ```suggestion
* **filesystem**: used primarily for server-side configurations outside the webpack build. Place the configuration file at the specified **path** and Irving will find your configuration.
```

## Customize Irving's express app
* **where to configure**: `server/customizeServer.js`
* **type**: function (run in sequence with all configured functions)
* **purpose**: Customize Irving's express server. Expects a function that will receive the express `app` for you to modify. Example:
```javascript
module.exports = function customizeServer(app) {
  app.get('/my-fun-route', routeHandler);
}
```

In order to properly integrate with Irving core, the cache service should return an object that looks like the following:

```javascript
const service = {
  client: null, // cache client object
  get: () => null, // get a cache key
  set: () => {}, // set a cache value
  del: () => null, // delete a key
  insert: () => null, // probably an alias for set
  remove: () => null, // probably an alias for del
  cached: () => {}, // set a value in the cache. Should be able to handle a promise as its value that will return the actual value to store in the cache
  close: () => {}, // close cache connection.
}
```

## Customize Irving's development server
* **where to configure**: `server/customizeServer.js`
* **type**: function (run in sequence with all configured functions)

> * **purpose**: Customize Irving's express server for development only (when you run the `irving dev` task). Expects a function that will receive the express `app` for you to modify. Example:

Reviewer: @todo add an example

## Customize Irving's production server
* **where to configure**: `server/customizeServer.js`
* **type**: function (run in sequence with all configured functions)

> * **purpose**: Customize Irving's express server for production only (when you run the `irving start` task). Expects a function that will receive the express `app` for you to modify. Example:

Reviewer: @todo add an example

## Customize domain redirects
* **where to configure**: `server/customizeRedirect.js`
* **type**: object
* **purpose**: Add custom redirect using the [Express Naked Redirect](https://www.npmjs.com/package/express-naked-redirect) package. You can supply the same config keys that the package accepts and Irving will use it to set up the redirect. Example:
```javascript
module.exports = {
    subDomain: 'cool',
    protocol: 'https',
}
```
The above config would set up the site to redirect from `http://interesting.site` into `https://cool.interesting.site`. It is important that the redirect is set up before activating this or you can run into issues.

## Default Redux State
* **where to configure**: `irving.config.js` in the `defaultState` key
* **type**: object
* **purpose**: Customize Irving's default redux state. Example:
```javascript
export default {
  form: {
    submitting: false,
    submitted: false,
    failed: false,
    validation: {},
    redirect: '',
  },
};

```

## Client-side environmental variable allowlist
* **where to configure**: `config/envAllowlist.js`
* **type**: array

> * **purpose**: Customize the allowlist of environmental variables that are passed to Irving's client-side javascript bundle. These env vars will be accessible via `window.__ENV__` or, if you need to access to the ENV object isomorphically, use the universal env utility via `import getEnv from '@irvingjs/core/utils/universalEnv';`

Reviewer: ```suggestion
* **purpose**: Customize the allowlist of environmental variables that are passed to Irving's client-side javascript bundle. These env vars will be accessible via `window.__ENV__` or, if you need access to the ENV object isomorphically, use the universal env utility via `import getEnv from '@irvingjs/core/utils/universalEnv';`
```

```javascript
const getEnvAllowlist = () => ([
  'MY_NEW_ENV_VAR',
]);

export default getEnvAllowlist;
```

## Force trailing slashes on URLs/links with the same hostname
* **where to configure**: `irving.config.js` in the `forceTrailingSlashes` key
* **type**: boolean
* **purpose**: Determines whether or not to add a trailing slash to any internal (same host) URL that doesn't currently have one.

## Server-side app template variables
* **where to configure**: `irving.config.js` in the `getAppTemplateVars` key
* **type**: object
* **purpose**: Customize template variables that are passed to Irving's server-rendered `app.ejs` file.
* **default values**: This config field contains some important default values that you may need to know how to handle.
  * `Wrapper` - App wrapper component. You may wrap this component in any other components or higher-order components (like a context provider).

>   * `head` - An object or function returning an object containing a variety of keys that will be rendered to the server-side template. These keys are treated differently than normal Irving config objects and can be provided with one of three values: an array, a string, or a function returning and array or string. All will eventually be concatenated into a single string and rendered as-is within the server-side HTML, meaning you should provide valid HTML to all of these fields.

Reviewer: ```suggestion
  * `head` - An object or function returning an object containing a variety of keys that will be rendered to the server-side template. These keys are treated differently than normal Irving config objects and can be provided with one of three values: an array, a string, or a function returning an array or string. All will eventually be concatenated into a single string and rendered as-is within the server-side HTML, meaning you should provide valid HTML to all of these fields.
```

```javascript
import myStateFieldReducer from './myStateFieldReducer';

export default {
  myStateField: myStateFieldReducer,
};
```

## Webpack Configuration
* **where to configure**: `config/webpack.config.js`
* **type**: object
* **purpose**: Customize Irving's internal webpack config for both server- and client-side builds. Example:
```javascript
// in config/getWebpackConfig.js
module.exports = (configs) => (
  // Map through both server and client configs.
  configs.map((config) => {
    config.module.rules.push({
      test: /\.scss$/,
      use: [
          'css-loader',
          'sass-loader',
      ],
    },);

    return config;
  })
);
```

## Server-side Webpack Configuration
* **where to configure**: `config/webpack.config.server.js`
* **type**: object
* **purpose**: Customize Irving's webpack configuration for only the server-side build.

## Client-side Webpack Configuration
* **where to configure**: `config/webpack.config.client.js`
* **type**: object

> * **purpose**: Customize Irving's webpack configuration for only the server-side build.

Reviewer: ```suggestion
* **purpose**: Customize Irving's webpack configuration for only the client-side build.
```
